syntax = "proto3";

package deepmind.reverb;

import "google/protobuf/duration.proto";
import "google/protobuf/timestamp.proto";
import "tensorflow/core/framework/tensor.proto";
import "tensorflow/core/protobuf/struct.proto";

// The actual data is stored in chunks. The data can be arbitrary tensors. We do
// not interpret the bytes data of the tensors on the server side. It is up to
// the client to compress the bytes blob within the tensors.
message ChunkData {
  // Unique identifier of the chunk.
  uint64 chunk_key = 1;

  // The timesteps within the episode that the chunk covers.
  SequenceRange sequence_range = 2;

  // Actual tensor data.
  repeated tensorflow.TensorProto data = 3;

  // True if delta encoding has been applied before compressing data.
  bool delta_encoded = 4;
}

// A range that specifies which items to slice out from a sequence of chunks.
// The length of all chunks must at least be `offset`+`length`.
message SliceRange {
  // Offset where the slice should start.
  int32 offset = 1;

  // Length of the slice. Can span multiple chunks.
  int32 length = 2;
}

message SequenceRange {
  // Globally unique identifier of the episode the sequence belongs to.
  uint64 episode_id = 1;

  // Index within the episode of the first timestep covered by the range.
  int32 start = 2;

  // Index within the episode of the last timestep covered by the range.
  // Must be >= start_index.
  int32 end = 3;
}

// A prioritized item is part of a table and references a chunk of
// data. Sampling happens based on the priority of items.
message PrioritizedItem {
  // Unique identifier of this item.
  uint64 key = 1;

  // Priority table that the item belongs to.
  string table = 2;

  // Sequence of chunks that is referenced by this item. Chunks in this list
  // will be batched together and are expected to have compatible shape and
  // dtype. Optional.
  repeated uint64 chunk_keys = 3;

  // Range that should be sliced out of the chunks in `chunk_keys`. This range
  // can span multiple chunks.
  SliceRange sequence_range = 4;

  // Priority used for sampling.
  double priority = 5;

  // The number of times the item has been sampled.
  int32 times_sampled = 6;

  // The time when the item was first inserted.
  google.protobuf.Timestamp inserted_at = 7;
}

// Used for updating an existing PrioritizedItem.
message KeyWithPriority {
  // Identifier of the PrioritizedItem.
  uint64 key = 1;

  // Priority used for sampling.
  double priority = 2;
}

message SampleInfo {
  // Item from that was sampled from the table.
  PrioritizedItem item = 1;

  // Probability that this item had at sampling time. Useful for importance
  // sampling.
  double probability = 2;

  // Number of items in the table at the time of the sample operation.
  int64 table_size = 3;

  // Priority of the item.
  double priority = 4;
}

// Metadata about the table, including (optional) data signature.
//
// These fields correspond to initialization arguments of the
// `Table` class, unless noted otherwise.
//
// Next ID: 11.
message TableInfo {
  // Table's name.
  string name = 8;

  // Sampler and remover metadata.
  KeyDistributionOptions sampler_options = 1;
  KeyDistributionOptions remover_options = 2;

  // Max size of the table.
  int64 max_size = 3;

  // Max number of times an element can be sampled before being
  // removed.
  int32 max_times_sampled = 4;

  // How data read/write is rate limited.
  RateLimiterInfo rate_limiter_info = 5;

  // Optional data signature for tensors stored in the table.  Note
  // that this data type is more flexible than we use.  For example,
  // we only store tensors (TensorSpecProto, TypeSpecProto) and not
  // any special data types (no NoneValue or other special fixed values).
  tensorflow.StructuredValue signature = 6;

  // Current size of table.
  int64 current_size = 7;

  // Number of episodes referenced by the items in the table.
  int64 num_episodes = 9;

  // Number of episodes once referenced by items in the table but no longer is.
  // The total number of episodes thus is `num_episodes + num_deleted_episodes`.
  int64 num_deleted_episodes = 10;
}

message RateLimiterCallStats {
  // The number of calls that are currently blocked.
  int64 pending = 1;

  // The total number of completed calls.
  int64 completed = 2;

  // Out of the completed calls, how many of these were blocked for any amount
  // of time.
  int64 limited = 3;

  // Total waiting time for all the COMPLETED calls.
  google.protobuf.Duration completed_wait_time = 4;

  // Total waiting time for all calls that are pending right now.
  google.protobuf.Duration pending_wait_time = 5;
}

message RateLimiterInfo {
  // The average number of times each item should be sampled during its
  // lifetime.
  double samples_per_insert = 1;

  // The minimum and maximum values the cursor is allowed to reach. The cursor
  // value is calculated as `insert_count * samples_per_insert -
  // sample_count`. If the value would go beyond these limits then the call is
  // blocked until it can proceed without violating the constraints.
  double min_diff = 2;
  double max_diff = 3;

  // The minimum number of inserts required before any sample operation.
  int64 min_size_to_sample = 4;

  // Stats regarding the limiting of insert calls.
  RateLimiterCallStats insert_stats = 5;

  // Stats regarding the limiting of sample calls.
  RateLimiterCallStats sample_stats = 6;
}

// Metadata about sampler or remover.  Describes its configuration.
message KeyDistributionOptions {
  message Prioritized {
    double priority_exponent = 1;
  }

  message Heap {
    bool min_heap = 1;
  }

  oneof distribution {
    bool fifo = 1;
    bool uniform = 2;
    Prioritized prioritized = 3;
    Heap heap = 4;
    bool lifo = 6;
  }
  reserved 5;
  bool is_deterministic = 7;
}

// Uint128 representation.  Can be used for unique identifiers.
message Uint128 {
  uint64 high = 1;
  uint64 low = 2;
}

// Representation of a timeout.  A value < 0 means never time out.
message Timeout {
  int64 milliseconds = 1;
}
